src/julia.h

```diff
@@ -1494,7 +1494,13 @@ STATIC_INLINE void jl_eh_restore_state(jl_handler_t *eh)
         locks->len = eh->locks_len;
     }
 #endif
-    JL_SIGATOMIC_END();
+    jl_get_ptls_states()->defer_signal = eh->defer_signal;
+    // This should be the last since this can trigger a safepoint
+    // that throws a SIGINT.
+    jl_gc_state_save_and_set(eh->gc_state);
+    if (eh->defer_signal == 0) {
+        jl_sigint_safepoint();
```

Comment:
The sigint safepoint is kind of avoidable (by reenable the gc safepoint) I don't think it's valid to avoid the GC safepoint in the gc state restore though.